File: 1335-maximum-candies-allocated-to-k-children/1335-maximum-candies-allocated-to-k-children.py
class Solution:
    def maximumCandies(self, candy: List[int], k: int) -> int:
        total_sum=sum(candy)

        if(total_sum<k):
            return 0
        l,r=1,total_sum//k
        ans=0

        while(l<=r):
            mid=(r+l)//2

            children=sum(pile//mid for pile in candy)

            if(children>=k):
                ans=mid
                l=mid+1
            else:
                r=mid-1
        return ans*1

            
        

File: 1335-maximum-candies-allocated-to-k-children/1335-maximum-candies-allocated-to-k-children.java
class Solution {
    public int maximumCandies(int[] candy, long k) {
        long sum=0;
        int ans=0;
        for(var x:candy) sum+=x;

        if(sum<k) return 0; //Candy is Less than children 'K'

        int l=1;
        int r=(int)(sum/k); //This is the maximum candy one can get

        while(l<=r){
            int mid=l +(r-l)/2;

            if(can(candy,mid,k)){
                ans=mid;
                l=mid+1; //Check larger value
            }else{
                r=mid-1; //Check smaller value
            }
        }
        return ans;

    }
    public boolean can(int[] candy, int pile, long k){
        long children=0;
        for(var x: candy) children +=x/pile;
        
        return children>=k;
    }
}